src/lem/eval.rs

```diff
@@ -9,63 +9,882 @@ pub(crate) fn eval_step() -> Func {
+                                                let (cont) = choose_let_cont(head, var, env, expanded, cont);
+                                                return (val, env, cont, ret)
+                                            }
+                                        };
+                                        return (expr, env, err, errctrl)
+                                    }
+                                };
+                                return (expr, env, err, errctrl)
+                            }
+                        };
+                        return (expr, env, err, errctrl)
+                    }
+                    Symbol("begin") => {
+                        let (arg1, more) = safe_uncons(rest);
+                        match more.tag {
+                            Expr::Nil => {
+                                return (arg1, env, cont, ret)
+                            }
+                        };
+                        let op2: Op2::Begin;
+                        let cont: Cont::Binop = hash4(op2, env, more, cont);
+                        return (arg1, env, cont, ret)
+                    }
+                    Symbol("eval") => {
+                        match rest.tag {
+                            Expr::Nil => {
+                                return (expr, env, err, errctrl)
+                            }
+                        };
+                        let (arg1, more) = safe_uncons(rest);
+                        match more.tag {
+                            Expr::Nil => {
+                                let op1: Op1::Eval;
+                                let cont: Cont::Unop = hash2(op1, cont);
+                                return (arg1, env, cont, ret)
+                            }
+                        };
+                        let op2: Op2::Eval;
+                        let cont: Cont::Binop = hash4(op2, env, more, cont);
+                        return (arg1, env, cont, ret)
+                    }
+                    Symbol("if") => {
+                        let (condition, more) = safe_uncons(rest);
+                        match more.tag {
+                            Expr::Nil => {
+                                return (condition, env, err, errctrl)
+                            }
+                        };
+                        let cont: Cont::If = hash2(more, cont);
+                        return (condition, env, cont, ret)
+                    }
+                    Symbol("current-env") => {
+                        match rest.tag {
+                            Expr::Nil => {
+                                return (env, env, cont, apply)
+                            }
+                        };
+                        return (expr, env, err, errctrl)
+                    }
+                };
+                // unops
+                let (op) = choose_unop(head);
```

Comment:
We don't need this (nor the incoming hack). We can simply match on the operators' symbols directly and get rid of `choose_unop`.

The "proper" way to do this in LEM with a `choose_unop` would be to implement support for some kind of `Option`. Then `choose_unop` would return `None` if no Op symbol is found.

Even if we decoupled tags and hashes (which I still think is a poor trade-off), the hash of a symbol is something we do not want to hardcode in a LEM because it's dependent on the field we are operating over.

So I would say we implement the version of eval that we had at some point, with flat cases of operation symbols listed in this match.